File: coordinates.f90
!   This is
!   https://github.com/LiuGangKingston/Nestable-coordinate-system-for-Tikz-circuits.git
!            Version 2.0
!   free for non-commercial use.
!   Please send us emails for any problems/suggestions/comments.
!   Please be advised that none of us accept any responsibility
!   for any consequences arising out of the usage of this
!   software, especially for damage.
!   For usage, please refer to the README file.
!   This code was written by
!        Gang Liu (gl.cell@outlook)
!                 (http://orcid.org/0000-0003-1575-9290)
!          and
!        Shiwei Huang (huang937@gmail.com)
!   Copyright (c) 2021 - 2022
!
!
module data
    implicit none
    integer,           parameter :: theinputtunnel  = 11
    integer,           parameter :: theoutputtunnel = 12
    integer,           parameter :: keylengthlimit  = 1000
    integer,           parameter :: linewidelimit   = 10 * keylengthlimit
    character (len=*), parameter :: xendkey         = 'xxx'
    character (len=*), parameter :: yendkey         = 'yyy'
    character (len=*), parameter :: pendkey         = 'ppp'
    character (len=*), parameter :: theinputfile    = 'in.dat'
    character (len=*), parameter :: macroset        = '\pgfmathsetmacro{'
    character (len=*), parameter :: macrosets       = '\pgfmathsetmacro{\'
    character (len=*), parameter :: spacing         = 'spacing'
    character (len=*), parameter :: additional      = ' + 0.0 }'
    character (len=*), parameter :: coordinate      = '\coordinate ('
contains


subroutine filegenerating(thefile, thetunnel, xxx, yyy, ppp, firstx, totalx, firsty, totaly)
    implicit none
    character (len=*) :: thefile, xxx, yyy, ppp
    character (len=1) :: firstx, firsty
    integer           :: thetunnel, totalx, totaly
    integer           :: firstxi, firstyi, ix, iy

    open(thetunnel, file=thefile)

    write(thetunnel,'(a)') '% https://github.com/LiuGangKingston/Nestable-coordinate-system-for-Tikz-circuits.git'
    write(thetunnel,'(a)') '% https://github.com/LiuGangKingston/Nestable-coordinate-system-for-Tikz-circuits.git'
    write(thetunnel,'(a)')
    write(thetunnel,'(a)')

    write(thetunnel,'(a,i2,a)') macroset//'\total'//xxx//'}{',totalx,'}'
    write(thetunnel,'(a,i2,a)') macroset//'\total'//yyy//'}{',totaly,'}'
    write(thetunnel,'(a)') macrosets//xxx//spacing//'}{1}'
    write(thetunnel,'(a)') macrosets//yyy//spacing//'}{1}'
    write(thetunnel,'(a)') macrosets//xxx//firstx//'}{-8}'
    write(thetunnel,'(a)') macrosets//yyy//firsty//'}{-8}'

    firstxi = ichar(firstx)
    firstyi = ichar(firsty)

    write(thetunnel,*)
    do ix = 1, totalx-1
        write(thetunnel,'(a)') macrosets//xxx//char(firstxi+ix)//'}{\'//xxx//char(firstxi-1+ix)//' + \'//xxx//spacing//additional
    end do

    write(thetunnel,*)
    do iy = 1, totaly-1
        write(thetunnel,'(a)') macrosets//yyy//char(firstyi+iy)//'}{\'//yyy//char(firstyi-1+iy)//' + \'//yyy//spacing//additional
    end do

    write(thetunnel,*)
    do ix = 1, totalx
    do iy = 1, totaly
        write(thetunnel,'(a)') coordinate//ppp//char(firstxi-1+ix)//char(firstyi-1+iy)//') at (\'//xxx//char(firstxi-1+ix)//&
                                                                                           &', \'//yyy//char(firstyi-1+iy)//');'
    end do
    end do

    write(thetunnel,*)
    write(thetunnel,'(a)') '%\gangprintcoordinateat{(0,0)}{The last coordinate values: }'//&
                             &'{($('//ppp//char(firstxi-1+totalx)//char(firstyi-1+totaly)//')$)}; '

    write(thetunnel,*)
    close(thetunnel)

    return
end subroutine filegenerating

end module


program coordinates
    use data
    implicit none
    integer :: linenumber = 0
    integer :: totalx, totaly, regenerate, ii, l, m, n
    character (len=linewidelimit) :: thekeyword, wordtmp, tmpword
    character (len=1) :: aaa, firstx, firsty

    write(*,*) 'If you want all the files: '
    write(*,*) '                "coordinates.tex" '
    write(*,*) '                "coorda.tex" '
    write(*,*) '                "coordb.tex" '
    write(*,*) '                 ...          '
    write(*,*) '                "coordz.tex" '
    write(*,*) '   to be (re-)generated, '
    write(*,*) 'please reply with 1. Otherwise, any other integer value.'
    read(*,*) regenerate
    if(regenerate.eq.1) then
        call filegenerating('coordinates.tex', theoutputtunnel, xendkey, yendkey, pendkey, 'a', 26, 'a', 26)
        do ii = 1, 26
           aaa = char(96+ii)
           call filegenerating('coord'//aaa//'.tex', theoutputtunnel, aaa//xendkey, aaa//yendkey, aaa//pendkey, 'a', 26, 'a', 26)
        end do
    end if

    linenumber = 0
    open(theinputtunnel, file=theinputfile)
    manyotherfiles: do

        linenumber = linenumber + 1
        wordtmp = ' '
        read(theinputtunnel, '(a)', end = 100)  wordtmp
        thekeyword = ' '
        thekeyword = adjustl(wordtmp)
        l = len_trim(thekeyword)
        if ((l.le.0) .or. (l.gt.keylengthlimit)) then
            write(*,*) 'The string in line number: ', linenumber, ' of the file '//theinputfile
            write(*,*) 'would be used as the KEY-WORD in this coordinate system.'
            write(*,*) 'Since either no such word or more than ', keylengthlimit, ' letters, then stopped.'
            stop
        end if

        do m = 1, l
             n = iachar(thekeyword(m:m))
             if ((n.le.96) .or. (n.ge.123)) then
               write(*,*) 'The string in line number: ', linenumber, ' of the file '//theinputfile
               write(*,*) 'would be used as the KEY-WORD in this coordinate system.'
               write(*,*) 'However only lower-case English letters are allowed.'
               write(*,*) 'Since other character: "'//thekeyword(m:m)//'" is found, then stopped.'
               stop
            end if
        end do

        linenumber = linenumber + 1
        wordtmp = ' '
        read(theinputtunnel, '(a)', end = 100)  wordtmp
        tmpword = ' '
        tmpword = adjustl(wordtmp)
        firstx = tmpword(1:1)
        if ((ichar(firstx).lt.97) .or. (ichar(firstx).gt.122)) then
            write(*,*) 'The first character in line number: ', linenumber, ' of the file '//theinputfile
            write(*,*) 'would be used as the first point in the x-direction in this coordinate system.'
            write(*,*) 'However it must be lower-case English letter.'
            write(*,*) 'Since it is "', firstx, '", then stopped.'
            stop
        end if

        linenumber = linenumber + 1
        read(theinputtunnel, *, end = 100)  totalx
        if ((totalx.lt.1) .or. (totalx.gt.(123-ichar(firstx)))) then
            write(*,*) 'The first integer in line number: ', linenumber, ' of the file '//theinputfile
            write(*,*) 'would be used as the total number of points in the x-direction in this coordinate system.'
            write(*,*) 'However it must be positive and no more than the number of letters between '//firstx//' and z.'
            write(*,*) 'Since it is ', totalx, ', then stopped.'
            stop
        end if

        linenumber = linenumber + 1
        wordtmp = ' '
        read(theinputtunnel, '(a)', end = 100)  wordtmp
        tmpword = ' '
        tmpword = adjustl(wordtmp)
        firsty = tmpword(1:1)
        if ((ichar(firsty).lt.97) .or. (ichar(firsty).gt.122)) then
            write(*,*) 'The first character in line number: ', linenumber, ' of the file '//theinputfile
            write(*,*) 'would be used as the first point in the y-direction in this coordinate system.'
            write(*,*) 'However it must be lower-case English letter.'
            write(*,*) 'Since it is "', firsty, '", then stopped.'
            stop
        end if

        linenumber = linenumber + 1
        read(theinputtunnel, *, end = 100)  totaly
        if ((totaly.lt.1) .or. (totaly.gt.(123-ichar(firsty)))) then
            write(*,*) 'The first integer in line number: ', linenumber, ' of the file '//theinputfile
            write(*,*) 'would be used as the total number of points in the y-direction in this coordinate system.'
            write(*,*) 'However it must be positive and no more than the number of letters between '//firsty//' and z.'
            write(*,*) 'Since it is ', totaly, ', then stopped.'
            stop
        end if


        call filegenerating('coord'//thekeyword(1:l)//'.tex', theoutputtunnel, thekeyword(1:l)//xendkey, &
                                                    &thekeyword(1:l)//yendkey, thekeyword(1:l)//pendkey, &
                                                                          & firstx, totalx, firsty, totaly)
    end do manyotherfiles

    100 stop

end program

File: coordinates.cpp
/*
!   This is
!   https://github.com/LiuGangKingston/Nestable-coordinate-system-for-Tikz-circuits.git
!            Version 2.0
!   free for non-commercial use.
!   Please send us emails for any problems/suggestions/comments.
!   Please be advised that none of us accept any responsibility
!   for any consequences arising out of the usage of this
!   software, especially for damage.
!   For usage, please refer to the README file.
!   This code was written by
!        Gang Liu (gl.cell@outlook)
!                 (http://orcid.org/0000-0003-1575-9290)
!          and
!        Shiwei Huang (huang937@gmail.com)
!   Copyright (c) 2021 - 2022
!
!
*/





#include <iostream>
#include <fstream>
#include <iomanip>
#include <vector>
#include <string>


namespace nestable_coordinate_system
{
    inline static const std::string xendkey      {"xxx"};
    inline static const std::string yendkey      {"yyy"};
    inline static const std::string pendkey      {"ppp"};
    inline static const std::string theinputfile {"in.dat"};
    inline static const std::string macroset     {"\\pgfmathsetmacro{"};
    inline static const std::string macrosets    {"\\pgfmathsetmacro{\\"};
    inline static const std::string spacing      {"spacing"};
    inline static const std::string additional   {" + 0.0 }"};
    inline static const std::string coordinate   {"\\coordinate ("};
    inline static const std::string whitespaces  {" \t"};
    inline static const std::string legalletters {"abcdefghijklmnopqrstuvwxyz"};


    inline void filegenerating(const std::string thefile, const std::string truexxx,
                               const std::string trueyyy, const std::string trueppp,
                               const char firstx,         const int totalx,
                               const char firsty,         const int totaly          ) noexcept
    {
        static std::ofstream nestablefile;
        int firstxi, firstyi, ix, iy;
        char txo, txp, tyo, typ;

        nestablefile.open(thefile);
        nestablefile << "% https://github.com/LiuGangKingston/Nestable-coordinate-system-for-Tikz-circuits.git" << std::endl;
        nestablefile << "% https://github.com/LiuGangKingston/Nestable-coordinate-system-for-Tikz-circuits.git" << std::endl;

        nestablefile << std::endl;
        nestablefile << std::endl;
        nestablefile << "%\\input{coordinates}" << std::endl;
        nestablefile << std::endl;
        nestablefile << std::endl;
        nestablefile << macroset << "\\total" << truexxx << "}{" << totalx << "}" << std::endl;
        nestablefile << macroset << "\\total" << trueyyy << "}{" << totaly << "}" << std::endl;
        nestablefile << macrosets << truexxx << spacing << "}{1}" << std::endl;
        nestablefile << macrosets << trueyyy << spacing << "}{1}" << std::endl;
        nestablefile << macrosets << truexxx << firstx << "}{-8}" << std::endl;
        nestablefile << macrosets << trueyyy << firsty << "}{-8}" << std::endl;

        nestablefile << std::endl;
        txo = firstx;
        for (ix = 1; ix < totalx; ix++)
        {
            txp = firstx + ix;
            nestablefile << macrosets << truexxx << txp << "}{\\" << truexxx << txo
                         << " + \\" << truexxx << spacing << additional << std::endl;
            txo = txp;
        }

        nestablefile << std::endl;
        tyo = firsty;
        for (iy = 1; iy < totaly; iy++)
        {
            typ = firsty + iy;
            nestablefile << macrosets << trueyyy << typ << "}{\\" << trueyyy << tyo
                         << " + \\" << trueyyy << spacing << additional << std::endl;
            tyo = typ;
        }

        nestablefile << std::endl;
        for (ix = 0; ix < totalx; ix++)
        {
            txp = firstx + ix;
            for (iy = 0; iy < totaly; iy++)
            {
                typ = firsty + iy;
                nestablefile << coordinate << trueppp << txp << typ << ") at (\\"
                             << truexxx << txp << ", \\" << trueyyy << typ << ");" << std::endl;
            }
        }

        nestablefile << std::endl;
        txp = firstx + totalx - 1;
        typ = firsty + totaly - 1;
        nestablefile << "%\\gangprintcoordinateat{(0,0)}{The last coordinate values: }{($("
                     << trueppp << txp << typ << ")$)}; " << std::endl;

        nestablefile << std::endl;
        nestablefile.close();

    }


}




void generatingdefaulfiles()
{
    using namespace nestable_coordinate_system;

    std::string aaa;
    int regenerate = 0;

    std::cout << "If you want all the following 27 files: " << std::endl;
    std::cout << "                \"coordinates.tex\" " << std::endl;
    std::cout << "                \"coorda.tex\" " << std::endl;
    std::cout << "                \"coordb.tex\" " << std::endl;
    std::cout << "                 ...          " << std::endl;
    std::cout << "                \"coordz.tex\" " << std::endl;
    std::cout << "   to be (re-)generated, " << std::endl;
    std::cout << "please input 1. Otherwise, any other integer value." << std::endl;
    std::cin  >> regenerate;
    if (regenerate == 1)
    {
        filegenerating("coordinates.tex", xendkey, yendkey, pendkey, 'a', 26, 'a', 26);

        for (int ii = 0; ii < 26; ii++)
        {
            aaa = 'a' + ii;
            filegenerating("coord" + aaa + ".tex", aaa + xendkey, aaa + yendkey, aaa + pendkey, 'a', 26, 'a', 26);
        }

        std::cout << "The above 27 coordinate system files were (re-)generated. "
                  << std::endl << std::endl;
    }
    else
    {
        std::cout << "None of the above 27 coordinate system files were touched. "
                  << std::endl << std::endl;
    }


}




void generatingadditionalfiles()
{
    using namespace nestable_coordinate_system;

    std::size_t left, right;
    std::string aline, thekeyword;
    char firstxpoint, firstypoint;
    int newfiles = 0, linenumber = 0, totalxpoints, totalypoints;

    std::ifstream theinput(theinputfile);
    if (theinput.fail())
    {
        std::cout << " Failed in openning the " << theinputfile << " file, then stopped. " << std::endl;
        exit(1);
    }

    while (true)
    {
        linenumber++;
        getline(theinput, aline);
        left = aline.find_first_not_of(whitespaces);
        if (left == std::string::npos)
        {
            std::cout << " No string at line #:" << linenumber
                      << " of the " << theinputfile << " file as KEYWORD, then stopped. " << std::endl;
            exit(1);
        }
        right = aline.find_last_not_of(whitespaces);
        thekeyword = aline.substr(left, right + 1);
        // middle = thekeyword.find_first_not_of(legalletters); // seeming not working.
        for (const auto& ch : thekeyword)
        {
            if (ch < 'a' || ch > 'z')
            {
                std::cout << "The string in line #: " << linenumber << " of the " << theinputfile << " file "
                          << "would be used as the KEY-WORD in this coordinate system. "
                          << "However only lower-case English letters are allowed. "
                          << "Since other character: " << ch <<" is found, then stopped. "
                          << std::endl;
                exit(1);
            }
        }


        linenumber++;
        getline(theinput, aline);
        left = aline.find_first_not_of(whitespaces);
        if (left == std::string::npos)
        {
            std::cout << " Nothing at line #:" << linenumber
                      << " of the " << theinputfile << " file as the first point in the x-direction, then stopped. " << std::endl;
            exit(1);
        }
        firstxpoint = aline[left];
        if (firstxpoint < 'a' || firstxpoint > 'z')
        {
            std::cout << "The first character in line #: " << linenumber << " of the " << theinputfile << " file "
                      << "would be used as the first point in the x-direction. "
                      << "However only lower-case English letters are allowed. "
                      << "Since it is " << firstxpoint <<", then stopped. "
                      << std::endl;
            exit(1);
        }


        linenumber++;
        totalxpoints = -1;
        // theinput >> totalxpoints;
        getline(theinput, aline);
        totalxpoints = std::stoi(aline);
        if (totalxpoints < 1 || totalxpoints > 'z' - firstxpoint + 1)
        {
            std::cout << "The number in line #: " << linenumber << " of the " << theinputfile << " file "
                      << "would be used as the total number of points in the x-direction. "
                      << "Since it is " << totalxpoints <<", too big/small, then stopped. "
                      << std::endl;
            exit(1);
        }


        linenumber++;
        getline(theinput, aline);
        left = aline.find_first_not_of(whitespaces);
        if (left == std::string::npos)
        {
            std::cout << " Nothing at line #:" << linenumber
                      << " of the " << theinputfile << " file as the first point in the y-direction, then stopped. " << std::endl;
            exit(1);
        }
        firstypoint = aline[left];
        if (firstypoint < 'a' || firstypoint > 'z')
        {
            std::cout << "The first character in line #: " << linenumber << " of the " << theinputfile << " file "
                      << "would be used as the first point in the y-direction. "
                      << "However only lower-case English letters are allowed. "
                      << "Since it is " << firstypoint <<", then stopped. "
                      << std::endl;
            exit(1);
        }


        linenumber++;
        totalypoints = -1;
        // theinput >> totalypoints;
        getline(theinput, aline);
        totalypoints = std::stoi(aline);
        if (totalypoints < 1 || totalypoints > 'z' - firstypoint + 1)
        {
            std::cout << "The number in line #: " << linenumber << " of the " << theinputfile << " file "
                      << "would be used as the total number of points in the y-direction. "
                      << "Since it is " << totalypoints <<", too big/small, then stopped. "
                      << std::endl;
            exit(1);
        }


        newfiles++;
        filegenerating("coord" + thekeyword + ".tex", thekeyword + xendkey, thekeyword + yendkey,
                       thekeyword + pendkey, firstxpoint, totalxpoints, firstypoint, totalypoints);
        std::cout << "New file #: " << newfiles << " was generated as: "
                  << "coord" + thekeyword + ".tex" << " based on: "
                  << firstxpoint << ' ' << totalxpoints << ' ' << firstypoint << ' ' << totalypoints  << '.'
                  << std::endl << std::endl;

    }

    theinput.close();


}




int main(void)
{
    generatingdefaulfiles();

    generatingadditionalfiles();

    return 0;
}


